File: Assets/Script/Lib/Agent/AudioToolRuntime.ts
// @preview-file off clear
import * as AudioGenerator from 'Agent/AudioGenerator';

export type AudioAgentToolName =
	| "generate_sfx"
	| "generate_music"
	| "generate_music_variation";

export type AudioToolProgress = AudioGenerator.GenerateSfxProgress | AudioGenerator.GenerateMusicProgress;

export interface AudioToolExecutionRequest {
	tool: AudioAgentToolName;
	params: Record<string, unknown>;
	workDir: string;
	isCancelled?: () => boolean;
	onProgress?: (progress: AudioToolProgress) => void;
}

export function isAudioAgentToolName(tool: string): tool is AudioAgentToolName {
	return tool === "generate_sfx"
		|| tool === "generate_music"
		|| tool === "generate_music_variation";
}

function hasParam(params: Record<string, unknown>, name: string): boolean {
	return params[name] !== undefined;
}

export function inferAudioToolNameFromParams(params: Record<string, unknown>): AudioAgentToolName | undefined {
	if (hasParam(params, "project") && hasParam(params, "path")) return "generate_music_variation";
	if (hasParam(params, "style") && hasParam(params, "path")) return "generate_music";
	if (hasParam(params, "type") && hasParam(params, "path")) return "generate_sfx";
	return undefined;
}

function stringParam(params: Record<string, unknown>, name: string): string {
	return typeof params[name] === "string" ? params[name] as string : "";
}

function optionalStringParam(params: Record<string, unknown>, name: string): string | undefined {
	return typeof params[name] === "string" ? params[name] as string : undefined;
}

function optionalNumberParam(params: Record<string, unknown>, name: string): number | undefined {
	return typeof params[name] === "number" ? params[name] as number : undefined;
}

function parseTonality(value: string | undefined): { key?: string; mode?: string; error?: string } {
	const text = (value ?? "auto").trim();
	if (text === "" || text.toLowerCase() === "auto") return {};
	const parts = text.split(" ").filter(part => part !== "");
	const key = parts[0]?.toUpperCase();
	const mode = parts.length > 1 ? parts.slice(1).join("_").toLowerCase() : undefined;
	const validKeys: string[] = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];
	const validModes: string[] = ["major", "minor", "pentatonic", "harmonic_minor", "dorian", "phrygian", "chromatic"];
	if (!key || validKeys.indexOf(key) < 0) return { error: `invalid tonality '${text}': expected a key such as D or F# dorian` };
	if (mode !== undefined && validModes.indexOf(mode) < 0) return { error: `invalid tonality '${text}': unknown mode '${mode}'` };
	return { key, mode };
}

export async function executeAudioTool(req: AudioToolExecutionRequest): Promise<AudioGenerator.GenerateSfxResult | AudioGenerator.GenerateMusicResult> {
	const params = req.params;
	if (req.tool === "generate_sfx") {
		return AudioGenerator.generateSfx({
			workDir: req.workDir,
			path: stringParam(params, "path"),
			type: stringParam(params, "type"),
			seed: optionalNumberParam(params, "seed"),
			volume: optionalNumberParam(params, "volume"),
			isCancelled: req.isCancelled,
			onProgress: req.onProgress,
		});
	}
	if (req.tool === "generate_music_variation") {
		return AudioGenerator.generateMusicVariation({
			workDir: req.workDir,
			project: stringParam(params, "project"),
			path: stringParam(params, "path"),
			seed: optionalNumberParam(params, "seed"),
			intensity: optionalNumberParam(params, "intensity"),
			variation: optionalNumberParam(params, "variation"),
			isCancelled: req.isCancelled,
			onProgress: req.onProgress,
		});
	}
	const tonality = parseTonality(optionalStringParam(params, "tonality"));
	const assetPack = optionalStringParam(params, "asset_pack") ?? "loop";
	if (tonality.error) return { success: false, path: stringParam(params, "path"), message: tonality.error };
	if (["loop", "adaptive", "cinematic", "full"].indexOf(assetPack) < 0) {
		return { success: false, path: stringParam(params, "path"), message: `invalid asset_pack '${assetPack}'` };
	}
	const cinematic = assetPack === "cinematic" || assetPack === "full";
	const adaptive = assetPack === "adaptive" || assetPack === "full";
	return AudioGenerator.generateMusic({
		workDir: req.workDir,
		path: stringParam(params, "path"),
		style: stringParam(params, "style"),
		seed: optionalNumberParam(params, "seed"),
		duration: optionalNumberParam(params, "duration"),
		bpm: optionalNumberParam(params, "bpm"),
		intensity: optionalNumberParam(params, "intensity"),
		key: tonality.key,
		mode: tonality.mode,
		stems: adaptive,
		introBars: cinematic ? 1 : 0,
		outroBars: cinematic ? 1 : 0,
		stinger: cinematic ? "both" : "none",
		exportMidi: assetPack === "full",
		isCancelled: req.isCancelled,
		onProgress: req.onProgress,
	});
}

File: Assets/Script/Lib/Agent/AudioToolRuntime.lua
-- [ts]: AudioToolRuntime.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__AsyncAwaiter = ____lualib.__TS__AsyncAwaiter -- 1
local __TS__Await = ____lualib.__TS__Await -- 1
local ____exports = {} -- 1
local AudioGenerator = require("Agent.AudioGenerator") -- 2
function ____exports.isAudioAgentToolName(tool) -- 19
	return tool == "generate_sfx" or tool == "generate_music" or tool == "generate_music_variation" -- 20
end -- 19
local function hasParam(params, name) -- 25
	return params[name] ~= nil -- 26
end -- 25
function ____exports.inferAudioToolNameFromParams(params) -- 29
	if hasParam(params, "project") and hasParam(params, "path") then -- 29
		return "generate_music_variation" -- 30
	end -- 30
	if hasParam(params, "style") and hasParam(params, "path") then -- 30
		return "generate_music" -- 31
	end -- 31
	if hasParam(params, "type") and hasParam(params, "path") then -- 31
		return "generate_sfx" -- 32
	end -- 32
	return nil -- 33
end -- 29
local function stringParam(params, name) -- 36
	return type(params[name]) == "string" and params[name] or "" -- 37
end -- 36
local function optionalStringParam(params, name) -- 40
	return type(params[name]) == "string" and params[name] or nil -- 41
end -- 40
local function optionalNumberParam(params, name) -- 44
	return type(params[name]) == "number" and params[name] or nil -- 45
end -- 44
local function optionalBooleanParam(params, name) -- 48
	local ____temp_0 -- 49
	if type(params[name]) == "boolean" then -- 49
		____temp_0 = params[name] -- 49
	else -- 49
		____temp_0 = nil -- 49
	end -- 49
	return ____temp_0 -- 49
end -- 48
function ____exports.executeAudioTool(req) -- 52
	return __TS__AsyncAwaiter(function(____awaiter_resolve) -- 52
		local params = req.params -- 53
		if req.tool == "generate_sfx" then -- 53
			return ____awaiter_resolve( -- 53
				nil, -- 53
				AudioGenerator.generateSfx({ -- 55
					workDir = req.workDir, -- 56
					path = stringParam(params, "path"), -- 57
					type = stringParam(params, "type"), -- 58
					seed = optionalNumberParam(params, "seed"), -- 59
					volume = optionalNumberParam(params, "volume"), -- 60
					isCancelled = req.isCancelled, -- 61
					onProgress = req.onProgress -- 62
				}) -- 62
			) -- 62
		end -- 62
		if req.tool == "generate_music_variation" then -- 62
			return ____awaiter_resolve( -- 62
				nil, -- 62
				AudioGenerator.generateMusicVariation({ -- 66
					workDir = req.workDir, -- 67
					project = stringParam(params, "project"), -- 68
					path = stringParam(params, "path"), -- 69
					seed = optionalNumberParam(params, "seed"), -- 70
					style = optionalStringParam(params, "style"), -- 71
					intensity = optionalNumberParam(params, "intensity"), -- 72
					variation = optionalNumberParam(params, "variation"), -- 73
					isCancelled = req.isCancelled, -- 74
					onProgress = req.onProgress -- 75
				}) -- 75
			) -- 75
		end -- 75
		return ____awaiter_resolve( -- 75
			nil, -- 75
			AudioGenerator.generateMusic({ -- 78
				workDir = req.workDir, -- 79
				path = stringParam(params, "path"), -- 80
				style = stringParam(params, "style"), -- 81
				seed = optionalNumberParam(params, "seed"), -- 82
				duration = optionalNumberParam(params, "duration"), -- 83
				bpm = optionalNumberParam(params, "bpm"), -- 84
				volume = optionalNumberParam(params, "volume"), -- 85
				intensity = optionalNumberParam(params, "intensity"), -- 86
				key = optionalStringParam(params, "key"), -- 87
				mode = optionalStringParam(params, "mode"), -- 88
				progression = optionalStringParam(params, "progression"), -- 89
				structure = optionalStringParam(params, "structure"), -- 90
				barsPerSection = optionalNumberParam(params, "bars_per_section"), -- 91
				melodyComplexity = optionalNumberParam(params, "melody_complexity"), -- 92
				rhythmComplexity = optionalNumberParam(params, "rhythm_complexity"), -- 93
				variation = optionalNumberParam(params, "variation"), -- 94
				leadInstrument = optionalStringParam(params, "lead_instrument"), -- 95
				bassInstrument = optionalStringParam(params, "bass_instrument"), -- 96
				harmonyInstrument = optionalStringParam(params, "harmony_instrument"), -- 97
				stereo = optionalBooleanParam(params, "stereo"), -- 98
				reverb = optionalNumberParam(params, "reverb"), -- 99
				delay = optionalNumberParam(params, "delay"), -- 100
				chorus = optionalNumberParam(params, "chorus"), -- 101
				distortion = optionalNumberParam(params, "distortion"), -- 102
				bitCrush = optionalNumberParam(params, "bit_crush"), -- 103
				lowPass = optionalNumberParam(params, "low_pass"), -- 104
				stems = optionalBooleanParam(params, "stems"), -- 105
				introBars = optionalNumberParam(params, "intro_bars"), -- 106
				outroBars = optionalNumberParam(params, "outro_bars"), -- 107
				stinger = optionalStringParam(params, "stinger"), -- 108
				exportMidi = optionalBooleanParam(params, "export_midi"), -- 109
				isCancelled = req.isCancelled, -- 110
				onProgress = req.onProgress -- 111
			}) -- 111
		) -- 111
	end) -- 111
end -- 52
return ____exports -- 52